core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java

```diff
@@ -382,10 +391,17 @@ public SyncStatus sync() throws AccessControlException, InvalidPathException {
         try {
           // we synced the path successfully
           // This shouldn't block because we checked job.isDone() earlier
-          if (job.get()) {
+          SyncResult result = job.get();
+          if (!result.isResultValid()) {
+            failedSyncPathCount++;
+          } else if (result.wasSyncPerformed()) {
             syncPathCount++;
           } else {
-            failedSyncPathCount++;
+            skippedSyncPathCount++;
+          }
+          if (result.isResultValid() && !result.wasSyncPerformed()) {
+            childOldestSkippedSync = childOldestSkippedSync == null ? result.getLastSyncTime()
```

Comment:
i'm just thinking aloud here.
if say the interval is 100s, the last synced timestamps are:
/a  0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for /a/b skipped. Then the timestamps look like the following:
/a  10
/a/b 10
/a/c 100
/a/d 100
/a/e 100
/a/f 100

and if we do a sync at timestamp 110, a metadata sync for /a will be triggered again, with only /a/b being sycned. After the operation, the timestamp looks like:
/a  110
/a/b 110
/a/c 100
/a/d 100
/a/e 100
/a/f 100

Is my understanding correct?

Reply:
My understanding is the same @elega so pls let me know if that's incorrect @tcrain :)

Reply:
When a sync happens on a directory, only the sync timestamp of the directory itself will be updated (including information if the sync was recursive). Then whenever a path checked it will check its parent sync time.

I think the reason for this (and why in the previous design indicated a failure if a child was more recent) was so that the sync cache would be updated only on the parent directory level and not for every child synced so that you would have less entires in the cache. And actually updating the children times individually would require a much more extensive redesign because the sync paths are not tracked in a way currently where they can track when their children finish (apart from the root sync path). I kept this idea of only updating the root sync directory instead of updating the individual children, but in reality I don't know what is best and would require benchmarks.

Taking this into account there are two choices for how to solve the problem of a child with a recent sync time.
1. What I have done in this PR, i.e. do not sync the child and only update the root sync path to the oldest skipped sync time.
2. Perform the sync on the child anyway and update the root sync path to the new time.

Lets go through your example, in the solution of 1 (note I am assuming every path here is a directory).

if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for /a/b skipped. Then the timestamps look like the following:
/a 10
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

and if we do a sync at timestamp 110, a metadata sync for /a will be triggered again, all children are synced. After the operation, the timestamp looks like (i.e. all paths have a sync time of 110):
/a 110
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

Now in the solution of 2.
if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0
the current timestamp is 100 and a sync will trigger with the sync for all children. Then the timestamps look like the following:
/a 100
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

and if we do a sync at timestamp 110 no sync is needed:
/a 110
/a/b 10
/a/c 0
/a/d 0
/a/e 0
/a/f 0

Here we see 2. works better than 1., but of course you can come up with an example that is the opposite where 1. works better

For example:

if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 0
/a/c 0
/a/d 0
/a/e 0
/a/f 0
now say at time 90 some of the children are synced individually. Then the timestamps look like the following:
/a 0
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

and if we do a sync at timestamp 100, a sync will only happen on /a/b, and /a will get updated to 90
/a 90
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

Now in the solution of 2.
if say the interval is 100s, the last synced timestamps are:
/a 0
/a/b 0
/a/c 0
/a/d 0
/a/e 0
/a/f 0
now say at time 90 some of the children are synced individually. Then the timestamps look like the following:
/a 0
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

and if we do a sync at timestamp 100, a a sync will happen on all children and /a will be updated to 100.
/a 100
/a/b 0
/a/c 90
/a/d 90
/a/e 90
/a/f 90

So basically what will work best depends on access pattern.

I chose solution 1 for two reasons:
- It is needed for cross cluster sync (there logical times are used so 1 is much more efficient)
- It gives more flexibility (see how only directories are skipped, and children that are files are still synced since the information is already available from listing the parent)

Reply:
thanks for the explanation i believe solution 1 is okay for now. I'm thinking in the future we can also probably record the time it took to do a metadata sync for a directory. With the help of such data, we can be adaptive between solution 1 or solution 2. For example, if you know that under a directory there's only one sub directory synced early and it only took half a second to finish. It is totally fine to go with solution 2 to sync everything because the cost is really low. On the other hand if we see a sub directory just synced or took a lot of time to sync, we can go with solution 1 to skip it. 

btw do you want to write down the example of your solution 1 as a comment in the code?

Reply:
@tcrain Thanks for the detailed examples! I overlooked the fact that children sync times are not updated but metadata sync checks parent sync times on the traversal. I'm fine with solution 1 and maybe in the future we define some scenarios to heuristically apply solution 2.